### PR TITLE
Formatting toolbar for RTE message composer.

### PR DESCRIPTION
**Pull request by aviraldg**

### CODE REVIEW

File: src/components/views/rooms/MessageComposer.js [hunk 1]
```diff
@@ -242,6 +258,17 @@ export default class MessageComposer extends React.Component {
             </div>;
         }
 
+
+        const {style, blockType} = this.state.selectionInfo;
+        const formatButtons = ["bold", "italic", "strike", "quote", "bullet", "numbullet"].map(
+            name => {
+                const active = style.includes(name) || blockType === name;
+                const suffix = active ? '-o-n' : '';
```
Comment (ara4n) [inline src/components/views/rooms/MessageComposer.js]: what does the `-o-n` naming scheme mean?

**Comment by ara4n** [comment]
lgtm, other than minor question. i haven't tested it though. is all the behaviour here hidden behind the labs flag?

File: src/components/views/rooms/MessageComposer.js [hunk 2]
```diff
@@ -48,9 +49,10 @@ export default class MessageComposer extends React.Component {
             inputState: {
                 style: [],
                 blockType: null,
-                isRichtextEnabled: true,
+                isRichtextEnabled: UserSettingsStore.getSyncedSetting('MessageComposerInput.isRichTextEnabled', true),
+                wordCount: 0,
             },
-            showFormatting: false,
+            showFormatting: UserSettingsStore.getSyncedSetting('MessageComposer.showFormatting', false),
```
Comment (ara4n) [inline src/components/views/rooms/MessageComposer.js]: have you tested this?

File: src/components/views/rooms/MessageComposerInput.js [hunk 3]
```diff
@@ -404,11 +404,13 @@ export default class MessageComposerInput extends React.Component {
...
-            let html = mdownToHtml(this.state.editorState.getCurrentContent().getPlainText());
-            this.setEditorState(this.createEditorState(enabled, RichText.HTMLtoContentState(html)));
+            const html = mdownToHtml(this.state.editorState.getCurrentContent().getPlainText());
+            const contentState = RichText.HTMLtoContentState(html);
+            this.setEditorState(this.createEditorState(enabled, contentState));
         } else {
-            let markdown = stateToMarkdown(this.state.editorState.getCurrentContent()),
-                contentState = ContentState.createFromText(markdown);
+            let markdown = stateToMarkdown(this.state.editorState.getCurrentContent());
+            markdown = markdown.substring(0, markdown.length - 1); // stateToMarkdown tacks on an extra newline (?!?)
```
Comment (ara4n) [inline src/components/views/rooms/MessageComposerInput.js]: as a sanity check ,can we only strip it if it's a \n?

**Comment by ara4n** [comment]
lgtm other than final comments (i commented on old commits, but the comments are still relevant)

**Comment by ara4n** [comment]
i'll merge in the morning if you can finish the last comments :)

File: src/components/views/rooms/MessageComposer.js (hunk 1 above)
Comment (aviraldg) [inline src/components/views/rooms/MessageComposer.js]: This was strangely the naming scheme used for the assets on Zeplin.

File: src/components/views/rooms/MessageComposer.js (hunk 2 above)
Comment (aviraldg) [inline src/components/views/rooms/MessageComposer.js]: What, whether it defaults to false or whether it persists correctly? It does seem to persist correctly.

**Pull request merged**